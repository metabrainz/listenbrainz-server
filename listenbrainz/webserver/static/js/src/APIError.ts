export default class APIError extends Error {
  status?: string;

  response?: Response;
}
