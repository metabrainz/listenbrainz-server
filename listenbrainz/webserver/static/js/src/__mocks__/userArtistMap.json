{
  "payload": {
    "artist_map": [
      { "artist_count": 50, "country": "GBR", "listen_count": 1846 },
      { "artist_count": 135, "country": "USA", "listen_count": 2705 },
      { "artist_count": 6, "country": "AUS", "listen_count": 534 },
      { "artist_count": 2, "country": "BEL", "listen_count": 227 },
      { "artist_count": 10, "country": "DEU", "listen_count": 216 },
      { "artist_count": 10, "country": "CAN", "listen_count": 152 },
      { "artist_count": 1, "country": "ARG", "listen_count": 77 },
      { "artist_count": 1, "country": "ESP", "listen_count": 73 },
      { "artist_count": 1, "country": "PRT", "listen_count": 61 },
      { "artist_count": 3, "country": "ISL", "listen_count": 63 },
      { "artist_count": 6, "country": "NLD", "listen_count": 63 },
      { "artist_count": 2, "country": "NOR", "listen_count": 28 },
      { "artist_count": 10, "country": "SWE", "listen_count": 48 },
      { "artist_count": 30, "country": "IND", "listen_count": 92 },
      { "artist_count": 5, "country": "IRL", "listen_count": 27 },
      { "artist_count": 5, "country": "FRA", "listen_count": 15 },
      { "artist_count": 1, "country": "URY", "listen_count": 4 },
      { "artist_count": 2, "country": "DNK", "listen_count": 3 },
      { "artist_count": 1, "country": "PAK", "listen_count": 2 },
      { "artist_count": 1, "country": "TUR", "listen_count": 1 },
      { "artist_count": 1, "country": "JAM", "listen_count": 1 },
      { "artist_count": 1, "country": "SGP", "listen_count": 1 },
      { "artist_count": 2, "country": "IDN", "listen_count": 2 },
      { "artist_count": 1, "country": "PHL", "listen_count": 1 },
      { "artist_count": 1, "country": "PRI", "listen_count": 1 },
      { "artist_count": 1, "country": "NZL", "listen_count": 1 },
      { "artist_count": 1, "country": "MYS", "listen_count": 1 }
    ],
    "from_ts": 1009843200,
    "last_updated": 1606535789,
    "range": "all_time",
    "to_ts": 1606499986,
    "user_id": "ishaanshah"
  }
}
