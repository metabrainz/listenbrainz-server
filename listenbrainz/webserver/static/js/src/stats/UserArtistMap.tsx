/* eslint-disable jsx-a11y/anchor-is-valid */
import * as React from "react";
import { faExclamationCircle, faLink } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { IconProp } from "@fortawesome/fontawesome-svg-core";

import APIService from "../APIService";
import Card from "../components/Card";
import Loader from "../components/Loader";
import Choropleth from "./Choropleth";

export type UserArtistMapProps = {
  range: UserStatsAPIRange;
  user: ListenBrainzUser;
  apiUrl: string;
};

export type UserArtistMapState = {
  data: UserArtistMapData;
  errorMessage?: string;
  graphContainerWidth?: number;
  hasError?: boolean;
  loading: boolean;
  selectedMetric: "artist" | "listen";
};

export default class UserArtistMap extends React.Component<
  UserArtistMapProps,
  UserArtistMapState
> {
  APIService: APIService;

  graphContainer: React.RefObject<HTMLDivElement>;

  rawData: UserArtistMapResponse;

  constructor(props: UserArtistMapProps) {
    super(props);
    this.APIService = new APIService(
      props.apiUrl || `${window.location.origin}/1`
    );

    this.state = {
      data: [],
      loading: false,
      errorMessage: "",
      hasError: false,
      selectedMetric: "artist",
    };

    this.graphContainer = React.createRef();

    this.rawData = {} as UserArtistMapResponse;
  }

  componentDidMount() {
    window.addEventListener("resize", this.handleResize);

    this.handleResize();
  }

  componentDidUpdate(prevProps: UserArtistMapProps) {
    const { range: prevRange } = prevProps;
    const { range: currRange } = this.props;
    if (prevRange !== currRange) {
      if (["week", "month", "year", "all_time"].indexOf(currRange) < 0) {
        this.setState({
          loading: false,
          hasError: true,
          errorMessage: `Invalid range: ${currRange}`,
        });
      } else {
        this.loadData();
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.handleResize);
  }

  changeSelectedMetric = (
    newSelectedMetric: "artist" | "listen",
    event?: React.MouseEvent<HTMLElement>
  ) => {
    if (event) {
      event.preventDefault();
    }

    this.setState({
      selectedMetric: newSelectedMetric,
      data: this.processData(this.rawData, newSelectedMetric),
    });
  };

  loadData = async (): Promise<void> => {
    const { selectedMetric } = this.state;
    this.setState({
      hasError: false,
      loading: true,
    });
    this.rawData = await this.getData();
    this.setState({
      data: this.processData(this.rawData, selectedMetric),
      loading: false,
    });
  };

  getData = async (): Promise<UserArtistMapResponse> => {
    const { range, user } = this.props;
    try {
      const data = await this.APIService.getUserArtistMap(user.name, range);
      return data;
    } catch (error) {
      if (error.response && error.response.status === 204) {
        this.setState({
          loading: false,
          hasError: true,
          errorMessage: "Statistics for the user have not been calculated",
        });
      } else {
        this.setState(() => {
          throw error;
        });
      }
    }
    return {} as UserArtistMapResponse;
  };

  processData = (
    data: UserArtistMapResponse,
    selectedMetric: "artist" | "listen"
  ): UserArtistMapData => {
    return data.payload.artist_map.map((country) => {
      return {
        id: country.country,
        value:
          selectedMetric === "artist"
            ? country.artist_count
            : country.listen_count,
      };
    });
  };

  handleResize = () => {
    this.setState({
      graphContainerWidth: this.graphContainer.current?.offsetWidth,
    });
  };

  render() {
    const {
      selectedMetric,
      data,
      errorMessage,
      graphContainerWidth,
      hasError,
      loading,
    } = this.state;

    return (
      <Card
        style={{
          marginTop: 20,
          minHeight: (graphContainerWidth || 1200) * 0.5,
        }}
        ref={this.graphContainer}
      >
        <div className="row">
          <div className="col-md-9 col-xs-6">
            <h3 className="capitalize-bold" style={{ marginLeft: 20 }}>
              Artist Origins
            </h3>
          </div>
          <div
            className="col-md-2 col-xs-4 text-right"
            style={{ marginTop: 20 }}
          >
            <span className="dropdown">
              <button
                className="dropdown-toggle btn-transparent capitalize-bold"
                data-toggle="dropdown"
                type="button"
              >
                {selectedMetric}s
                <span className="caret" />
              </button>
              <ul className="dropdown-menu" role="menu">
                <li>
                  <a
                    href=""
                    role="button"
                    onClick={(event) =>
                      this.changeSelectedMetric("listen", event)
                    }
                  >
                    Listens
                  </a>
                </li>
                <li>
                  <a
                    href=""
                    role="button"
                    onClick={(event) =>
                      this.changeSelectedMetric("artist", event)
                    }
                  >
                    Artists
                  </a>
                </li>
              </ul>
            </span>
          </div>
          <div className="col-md-1 col-xs-2 text-right">
            <h4 style={{ marginTop: 20 }}>
              <a href="#artist-origin">
                <FontAwesomeIcon
                  icon={faLink as IconProp}
                  size="sm"
                  color="#000000"
                  style={{ marginRight: 20 }}
                />
              </a>
            </h4>
          </div>
        </div>
        <Loader isLoading={loading}>
          {hasError && (
            <div
              className="flex-center"
              style={{
                minHeight: "inherit",
              }}
            >
              <span style={{ fontSize: 24 }} className="text-center">
                <FontAwesomeIcon icon={faExclamationCircle as IconProp} />{" "}
                {errorMessage}
              </span>
            </div>
          )}
          {!hasError && (
            <div className="row">
              <div className="col-xs-12">
                <Choropleth
                  data={data}
                  width={graphContainerWidth}
                  selectedMetric={selectedMetric}
                />
              </div>
            </div>
          )}
        </Loader>
      </Card>
    );
  }
}
