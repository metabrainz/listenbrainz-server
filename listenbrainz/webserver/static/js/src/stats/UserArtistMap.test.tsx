import * as React from "react";
import { mount, shallow } from "enzyme";

import UserArtistMap, { UserArtistMapProps } from "./UserArtistMap";
import APIError from "../APIError";
import * as userArtistMapResponse from "../__mocks__/userArtistMap.json";
import * as userArtistMapProcessedDataArtist from "../__mocks__/userArtistMapProcessDataArtist.json";
import * as userArtistMapProcessedDataListen from "../__mocks__/userArtistMapProcessDataListen.json";

const props: UserArtistMapProps = {
  user: {
    name: "foobar",
  },
  range: "week",
  apiUrl: "barfoo",
};

describe("UserArtistMap", () => {
  it("renders correctly", () => {
    const wrapper = shallow<UserArtistMap>(
      <UserArtistMap {...{ ...props, range: "all_time" }} />
    );

    wrapper.setState({
      selectedMetric: "artist",
      data: userArtistMapProcessedDataArtist,
      graphContainerWidth: 1200,
      loading: false,
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders corectly when range is invalid", () => {
    const wrapper = mount<UserArtistMap>(<UserArtistMap {...props} />);

    wrapper.setProps({ range: "invalid_range" as UserStatsAPIRange });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});

describe("componentDidMount", () => {
  it('adds event listener for "resize" event', () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "addEventListener");
    spy.mockImplementationOnce(() => {});
    instance.handleResize = jest.fn();
    instance.componentDidMount();

    expect(spy).toHaveBeenCalledWith("resize", instance.handleResize);
  });

  it('calls "handleResize" once', () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.handleResize = jest.fn();
    instance.componentDidMount();

    expect(instance.handleResize).toHaveBeenCalledTimes(1);
  });
});

describe("componentDidUpdate", () => {
  it("it sets correct state if range is incorrect", () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);

    wrapper.setProps({ range: "invalid_range" as UserStatsAPIRange });
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      loading: false,
      hasError: true,
      errorMessage: "Invalid range: invalid_range",
    });
  });

  it("calls loadData once if range is valid", () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.loadData = jest.fn();
    wrapper.setProps({ range: "month" });
    wrapper.update();

    expect(instance.loadData).toHaveBeenCalledTimes(1);
  });
});

describe("componentWillUnmount", () => {
  it('removes event listener for "resize" event', () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(window, "removeEventListener");
    spy.mockImplementationOnce(() => {});
    instance.handleResize = jest.fn();
    instance.componentWillUnmount();

    expect(spy).toHaveBeenCalledWith("resize", instance.handleResize);
  });
});

describe("getData", () => {
  it("calls getUserArtistMap with correct params", async () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserArtistMap");
    spy.mockImplementation(() =>
      Promise.resolve(userArtistMapResponse as UserArtistMapResponse)
    );
    const result = await instance.getData();

    expect(spy).toHaveBeenCalledWith("foobar", "week");
    expect(result).toEqual(userArtistMapResponse);
  });

  it("sets state correctly if data is not calculated", async () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserArtistMap");
    const noContentError = new APIError("NO CONTENT");
    noContentError.response = {
      status: 204,
    } as Response;
    spy.mockImplementation(() => Promise.reject(noContentError));
    await instance.getData();

    expect(wrapper.state()).toMatchObject({
      loading: false,
      hasError: true,
      errorMessage: "Statistics for the user have not been calculated",
    });
  });

  it("throws error", async () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    const spy = jest.spyOn(instance.APIService, "getUserArtistMap");
    const notFoundError = new APIError("NOT FOUND");
    notFoundError.response = {
      status: 404,
    } as Response;
    spy.mockImplementation(() => Promise.reject(notFoundError));

    await expect(instance.getData()).rejects.toThrow("NOT FOUND");
  });
});

describe("processData", () => {
  it("processes data correctly for all_time", () => {
    const wrapper = shallow<UserArtistMap>(
      <UserArtistMap {...{ ...props, range: "all_time" }} />
    );
    const instance = wrapper.instance();

    const result = instance.processData(
      userArtistMapResponse as UserArtistMapResponse,
      "artist"
    );

    expect(result).toEqual(userArtistMapProcessedDataArtist);
  });

  it("processes data correctly for listen", () => {
    const wrapper = shallow<UserArtistMap>(
      <UserArtistMap {...{ ...props, range: "all_time" }} />
    );
    const instance = wrapper.instance();

    const result = instance.processData(
      userArtistMapResponse as UserArtistMapResponse,
      "listen"
    );

    expect(result).toEqual(userArtistMapProcessedDataListen);
  });
});

describe("changeSelectedMetric", () => {
  it('sets state correctly for "artist"', () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.rawData = userArtistMapResponse as UserArtistMapResponse;

    instance.changeSelectedMetric("artist");
    expect(wrapper.state()).toMatchObject({
      data: userArtistMapProcessedDataArtist,
      selectedMetric: "artist",
    });
  });

  it('sets state correctly for "listen"', () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.rawData = userArtistMapResponse as UserArtistMapResponse;

    instance.changeSelectedMetric("listen");
    expect(wrapper.state()).toMatchObject({
      data: userArtistMapProcessedDataListen,
      selectedMetric: "listen",
    });
  });
});

describe("loadData", () => {
  it("calls getData once", async () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.getData = jest.fn();
    instance.processData = jest.fn();
    await instance.loadData();

    expect(instance.getData).toHaveBeenCalledTimes(1);
  });

  it("set state correctly", async () => {
    const wrapper = shallow<UserArtistMap>(<UserArtistMap {...props} />);
    const instance = wrapper.instance();

    instance.getData = jest
      .fn()
      .mockImplementationOnce(() => Promise.resolve(userArtistMapResponse));
    await instance.loadData();

    expect(instance.rawData).toMatchObject(userArtistMapResponse);

    expect(wrapper.state()).toMatchObject({
      data: userArtistMapProcessedDataArtist,
      loading: false,
    });
  });
});
