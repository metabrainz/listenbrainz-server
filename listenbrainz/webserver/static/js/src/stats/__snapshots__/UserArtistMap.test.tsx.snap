// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserArtistMap renders corectly when range is invalid 1`] = `
<UserArtistMap
  apiUrl="barfoo"
  range="invalid_range"
  user={
    Object {
      "name": "foobar",
    }
  }
>
  <ForwardRef
    style={
      Object {
        "marginTop": 20,
        "minHeight": 600,
      }
    }
  >
    <div
      style={
        Object {
          "background": "#FFFFFF",
          "border": "1px solid #EEEEEE",
          "borderRadius": "8px",
          "boxShadow": "0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20)",
          "boxSizing": "border-box",
          "height": "100%",
          "marginTop": 20,
          "minHeight": 600,
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="col-md-9 col-xs-6"
        >
          <h3
            className="capitalize-bold"
            style={
              Object {
                "marginLeft": 20,
              }
            }
          >
            Artist Origins
          </h3>
        </div>
        <div
          className="col-md-2 col-xs-4 text-right"
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <span
            className="dropdown"
          >
            <button
              className="dropdown-toggle btn-transparent capitalize-bold"
              data-toggle="dropdown"
              type="button"
            >
              artist
              s
              <span
                className="caret"
              />
            </button>
            <ul
              className="dropdown-menu"
              role="menu"
            >
              <li>
                <a
                  href=""
                  onClick={[Function]}
                  role="button"
                >
                  Listens
                </a>
              </li>
              <li>
                <a
                  href=""
                  onClick={[Function]}
                  role="button"
                >
                  Artists
                </a>
              </li>
            </ul>
          </span>
        </div>
        <div
          className="col-md-1 col-xs-2 text-right"
        >
          <h4
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <a
              href="#artist-origin"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                color="#000000"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      512,
                      512,
                      Array [],
                      "f0c1",
                      "M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z",
                    ],
                    "iconName": "link",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size="sm"
                spin={false}
                style={
                  Object {
                    "marginRight": 20,
                  }
                }
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-link fa-w-16 fa-sm "
                  color="#000000"
                  data-icon="link"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={
                    Object {
                      "marginRight": 20,
                    }
                  }
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </a>
          </h4>
        </div>
      </div>
      <Loader
        isLoading={false}
      >
        <div
          className="flex-center"
          style={
            Object {
              "minHeight": "inherit",
            }
          }
        >
          <span
            className="text-center"
            style={
              Object {
                "fontSize": 24,
              }
            }
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f06a",
                    "M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z",
                  ],
                  "iconName": "exclamation-circle",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-exclamation-circle fa-w-16 "
                data-icon="exclamation-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
             
            Invalid range: invalid_range
          </span>
        </div>
      </Loader>
    </div>
  </ForwardRef>
</UserArtistMap>
`;

exports[`UserArtistMap renders correctly 1`] = `
<ForwardRef
  style={
    Object {
      "marginTop": 20,
      "minHeight": 600,
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-md-9 col-xs-6"
    >
      <h3
        className="capitalize-bold"
        style={
          Object {
            "marginLeft": 20,
          }
        }
      >
        Artist Origins
      </h3>
    </div>
    <div
      className="col-md-2 col-xs-4 text-right"
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      <span
        className="dropdown"
      >
        <button
          className="dropdown-toggle btn-transparent capitalize-bold"
          data-toggle="dropdown"
          type="button"
        >
          artist
          s
          <span
            className="caret"
          />
        </button>
        <ul
          className="dropdown-menu"
          role="menu"
        >
          <li>
            <a
              href=""
              onClick={[Function]}
              role="button"
            >
              Listens
            </a>
          </li>
          <li>
            <a
              href=""
              onClick={[Function]}
              role="button"
            >
              Artists
            </a>
          </li>
        </ul>
      </span>
    </div>
    <div
      className="col-md-1 col-xs-2 text-right"
    >
      <h4
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <a
          href="#artist-origin"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            color="#000000"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f0c1",
                  "M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z",
                ],
                "iconName": "link",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size="sm"
            spin={false}
            style={
              Object {
                "marginRight": 20,
              }
            }
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </a>
      </h4>
    </div>
  </div>
  <Loader
    isLoading={false}
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12"
      >
        <CustomChoropleth
          data={
            Array [
              Object {
                "id": "GBR",
                "value": 50,
              },
              Object {
                "id": "USA",
                "value": 135,
              },
              Object {
                "id": "AUS",
                "value": 6,
              },
              Object {
                "id": "BEL",
                "value": 2,
              },
              Object {
                "id": "DEU",
                "value": 10,
              },
              Object {
                "id": "CAN",
                "value": 10,
              },
              Object {
                "id": "ARG",
                "value": 1,
              },
              Object {
                "id": "ESP",
                "value": 1,
              },
              Object {
                "id": "PRT",
                "value": 1,
              },
              Object {
                "id": "ISL",
                "value": 3,
              },
              Object {
                "id": "NLD",
                "value": 6,
              },
              Object {
                "id": "NOR",
                "value": 2,
              },
              Object {
                "id": "SWE",
                "value": 10,
              },
              Object {
                "id": "IND",
                "value": 30,
              },
              Object {
                "id": "IRL",
                "value": 5,
              },
              Object {
                "id": "FRA",
                "value": 5,
              },
              Object {
                "id": "URY",
                "value": 1,
              },
              Object {
                "id": "DNK",
                "value": 2,
              },
              Object {
                "id": "PAK",
                "value": 1,
              },
              Object {
                "id": "TUR",
                "value": 1,
              },
              Object {
                "id": "JAM",
                "value": 1,
              },
              Object {
                "id": "SGP",
                "value": 1,
              },
              Object {
                "id": "IDN",
                "value": 2,
              },
              Object {
                "id": "PHL",
                "value": 1,
              },
              Object {
                "id": "PRI",
                "value": 1,
              },
              Object {
                "id": "NZL",
                "value": 1,
              },
              Object {
                "id": "MYS",
                "value": 1,
              },
            ]
          }
          selectedMetric="artist"
          width={1200}
        />
      </div>
    </div>
  </Loader>
</ForwardRef>
`;
